### PR TITLE
Updating xUnit Performance Api and Performance Tests

### PR DESCRIPTION
**Pull request by jorive**
- This changes enable us to have a baseline and track performance changes in the product.
(The benchmarks and the way we measured has changed and we did not have an appropriate way to compare release/1.1.0 vs. release/2..0 vs. master)

### CODE REVIEW

**Comment by jorive** [comment]
@dotnet/rap-team PTAL

**Comment by jorive** [comment]
Currently I have one functional test failing on Windows, that I need to investigate:
```cmd
call CoreRun.exe xunit.console.netcore.exe System.Net.NetworkInformation.Functional.Tests.dll  -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapp1.0tests  -notrait category=OuterLoop -notrait category=failing -notrait category=nonwindowstests
xUnit.net console test runner (64-bit .NET Core)
Copyright (C) 2014 Outercurve Foundation.

Discovering: System.Net.NetworkInformation.Functional.Tests
Discovered:  System.Net.NetworkInformation.Functional.Tests
Starting:    System.Net.NetworkInformation.Functional.Tests
   System.Net.NetworkInformation.Tests.NetworkChangeTest.NetworkAddressChanged_AddRemove_Success [FAIL]
      System.IO.FileNotFoundException : Could not load file or assembly 'System.Threading.ThreadPool, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
      Stack Trace:
            at System.Net.NetworkInformation.NetworkChange.AddressChangeListener.StartHelper(NetworkAddressChangedEventHandler caller, Boolean captureContext, StartIPOptions startIPOptions)
            at System.Net.NetworkInformation.Tests.NetworkChangeTest.NetworkAddressChanged_AddRemove_Success()
Finished:    System.Net.NetworkInformation.Functional.Tests

=== TEST EXECUTION SUMMARY ===
   System.Net.NetworkInformation.Functional.Tests  Total: 47, Errors: 0, Failed: 1, Skipped: 0, Time: 0.312s
Finished running tests.  End time=12:20:41.30, Exit code = 1
```

File: src/System.IO.FileSystem.DriveInfo/tests/DriveInfo.Windows.Tests.cs [hunk 1]
```diff
@@ -199,7 +199,7 @@ public void VolumeLabelOnNetworkOrCdRom_Throws()
                 {
                     Exception e = Assert.ThrowsAny<Exception>(() => { adrive.VolumeLabel = null; });
                     Assert.True(
-                        e is UnauthorizedAccessException ||
```
Comment (mellinoe) [inline src/System.IO.FileSystem.DriveInfo/tests/DriveInfo.Windows.Tests.cs]: Are these non-perf test changes necessary? I know some of them are in shared files / base classes, but this one doesn't look necessary. We should try to touch as few files as possible, especially unrelated ones.

File: src/System.Collections/tests/Performance/Perf.Dictionary.cs [hunk 2]
```diff
@@ -20,8 +20,7 @@ public class Perf_Dictionary
             while (dict.Count < size)
             {
                 int key = rand.Next(500000, int.MaxValue);
-                if (!dict.ContainsKey(key))
-                    dict.Add(key, 0);
+                dict.TryAdd(key, 0);
```
Comment (mellinoe) [inline src/System.Collections/tests/Performance/Perf.Dictionary.cs]: This one is a bit unfortunate. We shouldn't have made this change in the first place (https://github.com/dotnet/corefx/pull/17201), because this is a perf test... I would say we should revert this everywhere (including master and release/2.0.0), but that is probably a non-starter this late in the game. This is okay as-is, I suppose.

File: src/System.IO.FileSystem.DriveInfo/tests/DriveInfo.Windows.Tests.cs (hunk 1 above)
Comment (mellinoe) [inline src/System.IO.FileSystem.DriveInfo/tests/DriveInfo.Windows.Tests.cs]: Are these non-perf test changes necessary? I know some of them are in shared files / base classes, but this one doesn't look necessary. We should try to touch as few files as possible, especially unrelated ones.

**Comment by danmosemsft** [comment]
@jorive can you describe how you produced these changes? did you basically xcopy all the perf test code from 2.0, remove anything using APi that didn't exist in 1.1, then pulled over all other test code that these changes relied on?

File: src/System.Collections/tests/Performance/Perf.Dictionary.cs (hunk 2 above)
Comment (jorive) [inline src/System.Collections/tests/Performance/Perf.Dictionary.cs]: I agree, and I think that should be part of a different PR, as this one is meant to set a common baseline to release/2.0.0 and master.

**Comment by jorive** [comment]
@danmosemsft, @mellinoe I tried to only update benchmarks, but that required me to update the common/tests projects they depend on, which subsequently forced me to update functional tests (as they depended directly on the same common projects).

The way I did these changes was by using a compare/diff tool between two corefx repos on release/1.1.0 and master branches. Next, I went through each benchmark project, inspect the csproj files, check for their dependencies to make sure we will be comparing the same projects, and pulled only non-release 2.0 changes (I did not bring benchmarks testing release 2.0 APIs). Then, build/run all tests and fix as necessary.

File: src/System.IO.FileSystem.DriveInfo/tests/DriveInfo.Windows.Tests.cs (hunk 1 above)
Comment (jorive) [inline src/System.IO.FileSystem.DriveInfo/tests/DriveInfo.Windows.Tests.cs]: I can remove this file. I initially made some changes that I reverted because they were not necessary, but my text editors are configured to trim those white spaces.

**Comment by mellinoe** [comment]
@jorive Could you rebase out the intermediate stuff when you have cleaned up all of the whitespace diffs?

**Comment by jorive** [comment]
@mellinoe Is it acceptable to just squash/merge the PR?

**Comment by mellinoe** [comment]
@jorive OK

**Comment by karelz** [comment]
@mellinoe what is left to move this PR forward?

**Comment by mellinoe** [comment]
@jorive The performance-related stuff all looks okay to me. We should try to prune out all of the unrelated test changes, and keep only the stuff that is necessary to get the perf tests in sync. The "changed line count" is a bit misleading because of all of the project.json and csproj files that needed changed, but there's still quite a lot of stuff in here that we could probably leave out.

**Comment by mellinoe** [comment]
@dotnet-bot Help

**Comment by mellinoe** [comment]
@dotnet-bot test outerloop Windows_NT Debug

test outerloop Ubuntu16.04 Release

**Comment by jorive** [comment]
@dotnet-bot test ci please

**Comment by jorive** [comment]
@mellinoe These are the errors on the two outerloops (Do we know if the outloop tests were clean?):
*Windows_NT Debug*
```
        System.Net.Http.WinHttpHandlerFunctional.Tests.WinHttpHandlerTest.SendAsync_SlowServerRespondsAfterDefaultReceiveTimeout_ThrowsHttpRequestException [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.AggregateException)
        Actual:   (No exception was thrown)
        Stack Trace:
           D:\j\workspace\outerloop_win---a7324842\src\System.Net.Http.WinHttpHandler\tests\FunctionalTests\WinHttpHandlerTest.cs(109,0): at System.Net.Http.WinHttpHandlerFunctional.Tests.WinHttpHandlerTest.SendAsync_SlowServerRespondsAfterDefaultReceiveTimeout_ThrowsHttpRequestException()
```
Related issue:
- https://github.com/dotnet/corefx/issues/11345

*Windows_NT Debug*
*Ubuntu16.04 Release*
```
        System.Net.Http.Functional.Tests.HttpClientTest.Timeout_SetTo60AndGetResponseFromServerWhichTakes40_Success [FAIL]
        Assert.Equal() Failure
        Expected: OK
        Actual:   ServiceUnavailable
        Stack Trace:
           D:\j\workspace\outerloop_win---a7324842\src\System.Net.Http\tests\FunctionalTests\HttpClientTest.cs(392,0): at System.Net.Http.Functional.Tests.HttpClientTest.Timeout_SetTo60AndGetResponseFromServerWhichTakes40_Success()
```
Related issues:
- https://github.com/dotnet/corefx/issues/8061
- https://github.com/dotnet/corefx/issues/12778

**Comment by mellinoe** [comment]
@jorive Did you run the tests locally?

I tried to start a couple of Jenkins runs against the base release/1.1.0 branch.

https://ci.dot.net/job/dotnet_corefx/job/release_1.1.0/job/outerloop_windows_nt_debug_prtest/
https://ci.dot.net/job/dotnet_corefx/job/release_1.1.0/job/outerloop_ubuntu16.04_release_prtest/

**Comment by mellinoe** [comment]
@jorive mentioned that the test failures repro'd for him on the release/1.1.0 branch without his changes.

The Jenkins runs have also started:

[Windows](https://ci.dot.net/job/dotnet_corefx/job/release_1.1.0/job/outerloop_windows_nt_debug_prtest/3/)
[Ubuntu](https://ci.dot.net/job/dotnet_corefx/job/release_1.1.0/job/outerloop_ubuntu16.04_release_prtest/3/)

**Comment by jorive** [comment]
@weshaggard PTAL

File: src/Common/tests/System/IO/Compression/LocalMemoryStream.cs [hunk 3]
```diff
@@ -2,7 +2,6 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System;
```
Comment (weshaggard) [inline src/Common/tests/System/IO/Compression/LocalMemoryStream.cs]: While these changes aren't wrong can you help me understand why all these clean-up type changes are included here?

**Comment by weshaggard** [comment]
I agree with @mellinoe that we should try and reduce this to only the necessary changes to avoid any issues in our release branch testing.

File: src/Common/tests/System/IO/Compression/LocalMemoryStream.cs (hunk 3 above)
Comment (jorive) [inline src/Common/tests/System/IO/Compression/LocalMemoryStream.cs]: I agree that some of those changes might not be necessary, but they already exist in master and it will make it easier, in the future, to port more changes with minimal effort (It is harder to go through every single line on those files than porting the projects where applicable and removing 2.* specific benchmarks, so I just moved the files).

**Comment by weshaggard** [comment]
@jorive OK as long as you have compared the functional test results before and after your change to ensure they match I can live with the test code changes to make diffing/merging easier.

**Comment by jorive** [comment]
@weshaggard The tests rate failure is the same or better with my changes, as I ported two test fixes.
